File: typescript/src/tools/mcpToolProvider.ts
import { AgentTool, AgentToolCallbacks, AgentToolResult, AgentTools, ToolResult } from "../utils/tool";
import { UIPayload } from "../utils/ui";
import { ConversationMessage } from "../types";

/** The advertised UI template URI: `_meta.ui.resourceUri`, or the OpenAI `openai/outputTemplate` alias. */
function uiResourceUri(meta: any): string | undefined {
  if (!meta || typeof meta !== "object") return undefined;
  const fromUi = meta.ui?.resourceUri;
  if (typeof fromUi === "string") return fromUi;
  const alias = meta["openai/outputTemplate"];
  return typeof alias === "string" ? alias : undefined;
}

/** Whether the model may be offered the tool. `_meta.ui.visibility` lists audiences; absent → both. */
function modelVisible(meta: any): boolean {
  const visibility = meta?.ui?.visibility;
  return Array.isArray(visibility) ? visibility.includes("model") : true;
}

/**
 * Configuration for a single MCP server connection.
 */
export interface MCPServerConfig {
  /** Transport type: stdio (spawn a local process) or sse (HTTP SSE endpoint) */
  type: "stdio" | "sse";
  /** stdio only: command to execute */
  command?: string;
  /** stdio only: arguments for the command */
  args?: string[];
  /** stdio only: environment variables for the child process */
  env?: Record<string, string>;
  /** sse only: full URL of the SSE endpoint */
  url?: string;
  /** sse only: extra HTTP headers */
  headers?: Record<string, string>;
}

/**
 * MCPToolProvider integrates one or more MCP (Model Context Protocol) servers
 * into the agent-squad tool system as a drop-in replacement for {@link AgentTools}.
 *
 * Use the async factory {@link MCPToolProvider.create} to build a provider.
 * It connects to all MCP servers before returning so that tool definitions are
 * available synchronously when the agent builds its API request:
 *
 * @example
 * ```typescript
 * const provider = await MCPToolProvider.create([
 *   { type: "stdio", command: "uvx", args: ["my-mcp-server"] },
 *   { type: "sse", url: "http://localhost:3000/sse" },
 * ]);
 *
 * const agent = new BedrockLLMAgent({
 *   name: "my-agent",
 *   description: "An agent with MCP tools",
 *   toolConfig: { tool: provider },
 * });
 *
 * // When done, clean up server connections:
 * await provider.disconnect();
 * ```
 *
 * The `@modelcontextprotocol/sdk` package must be installed separately:
 * ```
 * npm install @modelcontextprotocol/sdk
 * ```
 */
export class MCPToolProvider extends AgentTools {
  private servers: MCPServerConfig[];
  /** Connected MCP client instances, one per server */
  private clients: any[] = [];
  /** Map from MCP tool name → the client that owns it */
  private toolClientMap: Map<string, any> = new Map();
  /** Map from MCP tool name → its advertised UI resource URI (if any) */
  private toolUiMap: Map<string, string> = new Map();
  /** Per-client cache of fetched UI templates, keyed by resource URI */
  private templateCache: WeakMap<object, Map<string, { mimeType: string; body: string }>> =
    new WeakMap();
  private connected = false;

  constructor(servers: MCPServerConfig[], callbacks?: AgentToolCallbacks) {
    // Start with an empty tools array; populated lazily on first use
    super([], callbacks);
    this.servers = servers;
  }

  /**
   * Create a connected {@link MCPToolProvider}.
   *
   * Connects to all configured MCP servers and populates the internal tool list
   * before returning. Use this instead of `new MCPToolProvider(...)` so that
   * tool definitions are available when the agent builds its API request.
   *
   * @param servers - List of {@link MCPServerConfig} instances.
   * @param callbacks - Optional lifecycle hooks.
   * @returns A fully connected {@link MCPToolProvider}.
   */
  static async create(
    servers: MCPServerConfig[],
    callbacks?: AgentToolCallbacks
  ): Promise<MCPToolProvider> {
    const provider = new MCPToolProvider(servers, callbacks);
    await provider.ensureConnected();
    return provider;
  }

  // ---------------------------------------------------------------------------
  // Lazy connection
  // ---------------------------------------------------------------------------

  /**
   * Ensure all MCP servers are connected and the tool list is populated.
   * Safe to call multiple times — only runs once.
   */
  async ensureConnected(): Promise<void> {
    if (this.connected) return;

    let ClientClass: any;
    let StdioClientTransport: any;
    let SSEClientTransport: any;

    try {
      // @ts-ignore — optional peerDependency; not available in type-checking until installed
      const clientMod = await import("@modelcontextprotocol/sdk/client/index.js");
      ClientClass = clientMod.Client;
    } catch {
      throw new Error(
        "Install @modelcontextprotocol/sdk to use MCPToolProvider: npm install @modelcontextprotocol/sdk"
      );
    }

    try {
      // @ts-ignore — optional peerDependency
      const stdioMod = await import("@modelcontextprotocol/sdk/client/stdio.js");
      StdioClientTransport = stdioMod.StdioClientTransport;
    } catch {
      StdioClientTransport = null;
    }

    try {
      // @ts-ignore — optional peerDependency
      const sseMod = await import("@modelcontextprotocol/sdk/client/sse.js");
      SSEClientTransport = sseMod.SSEClientTransport;
    } catch {
      SSEClientTransport = null;
    }

    const allTools: AgentTool[] = [];

    for (const serverConfig of this.servers) {
      let transport: any;

      if (serverConfig.type === "stdio") {
        if (!StdioClientTransport) {
          throw new Error(
            "StdioClientTransport not available — check your @modelcontextprotocol/sdk installation"
          );
        }
        if (!serverConfig.command) {
          throw new Error(
            "MCPServerConfig with type 'stdio' requires a 'command' field"
          );
        }
        transport = new StdioClientTransport({
          command: serverConfig.command,
          args: serverConfig.args ?? [],
          env: serverConfig.env,
        });
      } else if (serverConfig.type === "sse") {
        if (!SSEClientTransport) {
          throw new Error(
            "SSEClientTransport not available — check your @modelcontextprotocol/sdk installation"
          );
        }
        if (!serverConfig.url) {
          throw new Error(
            "MCPServerConfig with type 'sse' requires a 'url' field"
          );
        }
        transport = new SSEClientTransport(new URL(serverConfig.url), {
          headers: serverConfig.headers ?? {},
        });
      } else {
        throw new Error(
          `Unsupported MCPServerConfig type: ${(serverConfig as any).type}`
        );
      }

      const client = new ClientClass(
        { name: "agent-squad-mcp-client", version: "1.0.0" },
        { capabilities: {} }
      );

      await client.connect(transport);
      this.clients.push(client);

      const { tools: mcpTools } = await client.listTools();

      for (const mcpTool of mcpTools) {
        const toolName = mcpTool.name;
        this.toolClientMap.set(toolName, client);

        const meta = mcpTool._meta;
        const ui = uiResourceUri(meta);
        if (ui) this.toolUiMap.set(toolName, ui);

        const schema = mcpTool.inputSchema ?? { type: "object", properties: {} };
        const properties: Record<string, any> = schema.properties ?? {};
        const required: string[] = schema.required ?? [];

        // A real func: the gatherer invokes it through the base tool loop, which lets a UI-aware
        // consumer (GroundedAgent) capture the ToolResult it returns — including any widget.
        const agentTool = new AgentTool({
          name: toolName,
          description: mcpTool.description ?? `MCP tool: ${toolName}`,
          properties,
          required,
          func: async (input: any) => this.callMCPTool(toolName, input),
        });

        // Store the raw MCP input schema so format methods can pass it through
        // unchanged (it is already valid JSON Schema).
        (agentTool as any)._mcpInputSchema = schema;

        // App-only tools stay callable (in toolClientMap) but are never advertised to the model.
        if (modelVisible(meta)) allTools.push(agentTool);
      }
    }

    this.tools = allTools;
    this.connected = true;
  }

  // ---------------------------------------------------------------------------
  // Format overrides — pass MCP inputSchema through directly
  // ---------------------------------------------------------------------------

  /**
   * Returns tool definitions in Amazon Bedrock Converse format.
   * Triggers lazy MCP connection on first call.
   */
  async toBedrockFormat(): Promise<any[]> {
    await this.ensureConnected();
    return this.tools.map((tool) => ({
      toolSpec: {
        name: tool.name,
        description: tool.description,
        inputSchema: {
          json: (tool as any)._mcpInputSchema ?? {
            type: "object",
            properties: tool.properties,
            required: tool.required,
          },
        },
      },
    }));
  }

  /**
   * Returns tool definitions in Anthropic Claude format.
   * Triggers lazy MCP connection on first call.
   */
  async toAnthropicFormat(): Promise<any[]> {
    await this.ensureConnected();
    return this.tools.map((tool) => ({
      name: tool.name,
      description: tool.description,
      input_schema: (tool as any)._mcpInputSchema ?? {
        type: "object",
        properties: tool.properties,
        required: tool.required,
      },
    }));
  }

  /**
   * Returns tool definitions in OpenAI Chat Completions format.
   * Triggers lazy MCP connection on first call.
   */
  async toOpenAIFormat(): Promise<any[]> {
    await this.ensureConnected();
    return this.tools.map((tool) => ({
      type: "function",
      function: {
        name: tool.name,
        description: tool.description,
        parameters: (tool as any)._mcpInputSchema ?? {
          type: "object",
          properties: tool.properties,
          required: tool.required,
        },
      },
    }));
  }

  // ---------------------------------------------------------------------------
  // toolHandler override — routes to MCP instead of local funcs
  // ---------------------------------------------------------------------------

  /**
   * Runs each tool-use block by invoking the tool's `func` (our MCP call, returning a
   * {@link ToolResult}) with the full input, then routes only its text to the model. Calling `func`
   * is what lets a UI-aware consumer such as `GroundedAgent` capture the widget-carrying result.
   *
   * We invoke `func` directly (rather than delegating to the base loop) so the tool receives its
   * full input — the base `processTool` would unwrap a `messages` key and drop the other arguments.
   */
  override async toolHandler(
    response: any,
    getToolUseBlock: (block: any) => any,
    getToolName: (toolUseBlock: any) => string,
    getToolId: (toolUseBlock: any) => string,
    getInputData: (toolUseBlock: any) => any
  ): Promise<AgentToolResult[]> {
    await this.ensureConnected();

    if (!response.content) {
      throw new Error("No content blocks in response");
    }

    const toolResults: AgentToolResult[] = [];

    for (const block of response.content) {
      const toolUseBlock = getToolUseBlock(block);
      if (!toolUseBlock) continue;

      const toolName = getToolName(toolUseBlock);
      const toolId = getToolId(toolUseBlock);
      const inputData = getInputData(toolUseBlock);

      // Model-visible tools carry a real func (captured by GroundedAgent); app-only / unknown tools
      // aren't advertised, so fall back to a direct MCP call.
      // Only model-visible tools (in this.tools) are executable via the model loop. An app-only or
      // unknown name is rejected here — an app-only MCP tool must never be invocable by the model.
      const tool = this.tools.find((t) => t.name === toolName);
      const result = tool
        ? await tool.func(inputData)
        : new ToolResult(`MCP tool '${toolName}' not found`);

      const content =
        result instanceof ToolResult
          ? result.content || JSON.stringify(result.structuredContent)
          : result;
      toolResults.push(new AgentToolResult(toolId, content));
    }

    return toolResults;
  }

  // ---------------------------------------------------------------------------
  // Internal helpers
  // ---------------------------------------------------------------------------

  private async callMCPTool(toolName: string, inputData: any): Promise<ToolResult> {
    const client = this.toolClientMap.get(toolName);
    if (!client) {
      return new ToolResult(`MCP tool '${toolName}' not found`);
    }

    try {
      const mcpResult = await client.callTool({
        name: toolName,
        arguments: inputData ?? {},
      });

      const text = this.extractTextFromContent(mcpResult.content);
      if (mcpResult.isError) {
        return new ToolResult(`Error from MCP tool '${toolName}': ${text}`);
      }

      const structured = mcpResult.structuredContent ?? {};
      let ui: UIPayload | undefined;
      const resourceUri = this.toolUiMap.get(toolName);
      if (resourceUri) {
        const template = await this.readTemplate(client, resourceUri);
        if (template) {
          ui = {
            resourceUri,
            mimeType: template.mimeType,
            template: template.body,
            structuredContent: structured,
            meta: mcpResult._meta,
          };
        }
      }
      return new ToolResult(text, structured, ui);
    } catch (error: any) {
      return new ToolResult(
        `Error calling MCP tool '${toolName}': ${error?.message ?? String(error)}`
      );
    }
  }

  /** Fetch (and cache, per client) a UI template resource. A resource is text or a base64 blob. */
  private async readTemplate(
    client: any,
    resourceUri: string
  ): Promise<{ mimeType: string; body: string } | undefined> {
    const cached = this.templateCache.get(client);
    if (cached?.has(resourceUri)) return cached.get(resourceUri);

    let template: { mimeType: string; body: string } | undefined;
    try {
      const res = await client.readResource({ uri: resourceUri });
      const first = res?.contents?.[0];
      if (first) {
        const mimeType: string = first.mimeType ?? "text/html;profile=mcp-app";
        let body: string | undefined = typeof first.text === "string" ? first.text : undefined;
        if (body === undefined && typeof first.blob === "string") {
          body = Buffer.from(first.blob, "base64").toString("utf-8");
        }
        if (body !== undefined) template = { mimeType, body };
      }
    } catch {
      template = undefined;
    }

    if (template) {
      const byUri = cached ?? new Map<string, { mimeType: string; body: string }>();
      byUri.set(resourceUri, template);
      this.templateCache.set(client, byUri);
    }
    return template;
  }

  private extractTextFromContent(content: any): string {
    if (!content) return "";
    if (typeof content === "string") return content;
    if (Array.isArray(content)) {
      return content
        .map((item: any) => {
          if (typeof item === "string") return item;
          if (item?.type === "text") return item.text ?? "";
          if (item?.text) return item.text;
          return JSON.stringify(item);
        })
        .join("\n");
    }
    if (typeof content === "object" && content.text) return content.text;
    return JSON.stringify(content);
  }

  /**
   * Disconnect all MCP clients. Call this when the agent is no longer needed
   * to cleanly shut down stdio child processes or SSE connections.
   */
  async disconnect(): Promise<void> {
    for (const client of this.clients) {
      try {
        await client.close();
      } catch {
        // ignore errors during cleanup
      }
    }
    this.clients = [];
    this.toolClientMap.clear();
    this.toolUiMap.clear();
    this.templateCache = new WeakMap();
    this.tools = [];
    this.connected = false;
  }
}

// Re-export for convenience
export type { ConversationMessage };

File: python/src/agent_squad/tools/mcp_tool_provider.py
"""
MCPToolProvider — drop-in AgentTools subclass that exposes MCP server tools.

Use the async factory :meth:`MCPToolProvider.create` to build a provider.
This connects to all MCP servers upfront so that tool definitions are
available synchronously when the agent builds its API request::

    from agent_squad.tools import MCPToolProvider, MCPServerConfig
    from agent_squad.agents import BedrockLLMAgent, BedrockLLMAgentOptions

    provider = await MCPToolProvider.create([
        MCPServerConfig(type="stdio", command="uvx", args=["my-mcp-server"]),
        MCPServerConfig(type="sse", url="http://localhost:3000/sse"),
    ])

    agent = BedrockLLMAgent(BedrockLLMAgentOptions(
        name="my-agent",
        description="An agent with MCP tools",
        tool_config={"tool": provider}
    ))

    # When done, clean up server connections:
    await provider.disconnect()

Requires the ``mcp`` extra::

    pip install agent-squad[mcp]
"""

from __future__ import annotations

import base64
import json

try:
    from mcp import ClientSession
    from mcp.client.stdio import stdio_client, StdioServerParameters
    from mcp.client.sse import sse_client
except ImportError as exc:
    raise ImportError(
        "MCPToolProvider requires the 'mcp' package. "
        "Install it with: pip install agent-squad[mcp]"
    ) from exc

from pydantic import AnyUrl  # mcp depends on pydantic, so it's available whenever the import above succeeds

from dataclasses import dataclass, field
from typing import Any, Optional

from agent_squad.types import AgentProviderType, ConversationMessage, ParticipantRole
from agent_squad.utils.tool import AgentTools, AgentToolCallbacks, ToolResult
from agent_squad.utils.ui import UIPayload


@dataclass
class MCPServerConfig:
    """Configuration for a single MCP server.

    For stdio transport set ``type="stdio"`` and provide ``command`` / ``args`` / ``env``.
    For SSE/HTTP transport set ``type="sse"`` and provide ``url`` / ``headers``.

    Attributes:
        type: Transport type — ``"stdio"`` or ``"sse"``.
        command: Executable to launch (stdio only).
        args: Command-line arguments (stdio only).
        env: Environment variables to pass to the subprocess (stdio only).
        url: SSE endpoint URL (sse only).
        headers: HTTP headers to send with the SSE connection (sse only).
    """

    type: str  # "stdio" or "sse"
    command: Optional[str] = None
    args: list[str] = field(default_factory=list)
    env: Optional[dict[str, str]] = None
    url: Optional[str] = None
    headers: Optional[dict[str, str]] = None


def _meta_dict(mcp_tool: Any) -> Optional[dict[str, Any]]:
    """The tool's ``_meta`` (MCP Apps UI metadata); the SDK exposes it as ``.meta``."""
    return getattr(mcp_tool, "meta", None) or getattr(mcp_tool, "_meta", None)


def _ui_resource_uri(meta: Optional[dict[str, Any]]) -> Optional[str]:
    """The advertised UI template URI: ``_meta.ui.resourceUri``, or the OpenAI
    ``openai/outputTemplate`` alias. Tolerant of absent/wrong-typed fields."""
    if not isinstance(meta, dict):
        return None
    ui = meta.get("ui")
    if isinstance(ui, dict) and isinstance(ui.get("resourceUri"), str):
        return ui["resourceUri"]
    alias = meta.get("openai/outputTemplate")
    return alias if isinstance(alias, str) else None


def _model_visible(meta: Optional[dict[str, Any]]) -> bool:
    """Whether the model may be offered the tool. ``_meta.ui.visibility`` lists audiences
    (``model`` / ``app``); absent means both (the MCP default)."""
    if not isinstance(meta, dict):
        return True
    ui = meta.get("ui")
    if not isinstance(ui, dict):
        return True
    visibility = ui.get("visibility")
    if not isinstance(visibility, list):
        return True
    return "model" in visibility


@dataclass
class _MCPToolEntry:
    """Per-tool state: owning session, the raw MCP tool, its advertised UI resource URI (if any),
    and whether the model may see it (app-only tools stay callable but unadvertised)."""

    session: Any
    tool: Any
    ui: Optional[str] = None
    model_visible: bool = True


class MCPToolProvider(AgentTools):
    """AgentTools subclass that proxies tools from one or more MCP servers.

    Use the async class method :meth:`create` to build a provider.  It connects
    to all configured MCP servers and populates the tool list before returning,
    so tool definitions are available synchronously when the agent builds its
    API request::

        provider = await MCPToolProvider.create([
            MCPServerConfig(type="stdio", command="uvx", args=["my-server"]),
            MCPServerConfig(type="sse", url="http://localhost:3000/sse"),
        ])
        tool_config={"tool": provider}

    Call :meth:`disconnect` when the provider is no longer needed to cleanly
    shut down stdio child processes or SSE connections.

    Args:
        servers: List of :class:`MCPServerConfig` describing the MCP servers to
            connect to.
        callbacks: Optional :class:`~agent_squad.utils.tool.AgentToolCallbacks`
            instance for lifecycle hooks.
    """

    def __init__(
        self,
        servers: list[MCPServerConfig],
        callbacks: Optional[AgentToolCallbacks] = None,
    ) -> None:
        super().__init__(tools=[], callbacks=callbacks)

        self._servers = servers
        # Maps tool_name → _MCPToolEntry
        self._tool_map: dict[str, _MCPToolEntry] = {}
        # Caches fetched UI templates: (session id, resourceUri) → (mime_type, body).
        # Keyed by session too, so two servers advertising the same URI can't collide.
        self._template_cache: dict[tuple[int, str], tuple[str, str]] = {}
        self._connected = False
        # Keep hold of context-manager stacks so we can exit them on disconnect
        self._cm_stack: list[Any] = []
        self._sessions: list[Any] = []

    @classmethod
    async def create(
        cls,
        servers: list[MCPServerConfig],
        callbacks: Optional[AgentToolCallbacks] = None,
    ) -> "MCPToolProvider":
        """Create a connected :class:`MCPToolProvider`.

        Connects to all configured MCP servers and populates the internal tool
        map before returning.  Use this instead of constructing the class
        directly so that tool definitions are available when the agent builds
        its API request.

        Args:
            servers: List of :class:`MCPServerConfig` instances.
            callbacks: Optional lifecycle hooks.

        Returns:
            A fully connected :class:`MCPToolProvider` instance.
        """
        provider = cls(servers, callbacks)
        await provider._ensure_connected()
        return provider

    # ------------------------------------------------------------------
    # Internal connection management
    # ------------------------------------------------------------------

    async def _ensure_connected(self) -> None:
        """Lazily connect to all configured servers and cache their tools."""
        if self._connected:
            return

        for server_cfg in self._servers:
            if server_cfg.type == "stdio":
                if not server_cfg.command:
                    raise ValueError("MCPServerConfig with type='stdio' requires a 'command'")
                params = StdioServerParameters(
                    command=server_cfg.command,
                    args=server_cfg.args or [],
                    env=server_cfg.env,
                )
                cm = stdio_client(params)
            elif server_cfg.type == "sse":
                if not server_cfg.url:
                    raise ValueError("MCPServerConfig with type='sse' requires a 'url'")
                cm = sse_client(server_cfg.url, headers=server_cfg.headers or {})
            else:
                raise ValueError(
                    f"Unsupported MCPServerConfig type: '{server_cfg.type}'. "
                    "Use 'stdio' or 'sse'."
                )

            read, write = await cm.__aenter__()
            self._cm_stack.append(cm)

            session = ClientSession(read, write)
            await session.__aenter__()
            self._sessions.append(session)
            await session.initialize()

            tools_result = await session.list_tools()
            for mcp_tool in tools_result.tools:
                meta = _meta_dict(mcp_tool)
                self._tool_map[mcp_tool.name] = _MCPToolEntry(
                    session=session,
                    tool=mcp_tool,
                    ui=_ui_resource_uri(meta),
                    model_visible=_model_visible(meta),
                )

        self._connected = True

    async def disconnect(self) -> None:
        """Disconnect from all MCP servers and release resources.

        Closes all client sessions and transport context managers.  After
        calling this method the provider must not be used again.
        """
        for session in self._sessions:
            try:
                await session.__aexit__(None, None, None)
            except Exception:  # noqa: BLE001
                pass
        self._sessions = []

        for cm in self._cm_stack:
            try:
                await cm.__aexit__(None, None, None)
            except Exception:  # noqa: BLE001
                pass
        self._cm_stack = []

        self._tool_map = {}
        self._template_cache = {}
        self._connected = False

    # ------------------------------------------------------------------
    # AgentTools interface
    # ------------------------------------------------------------------

    async def tool_handler(
        self,
        provider_type: str,
        response: Any,
        _conversation: list[dict[str, Any]],
        agent_info: Optional[dict[str, Any]] = None,
    ) -> Any:
        """Execute tool calls found in *response* and return results.

        Compatible with Bedrock, Anthropic, and OpenAI response shapes.
        """
        await self._ensure_connected()

        if not response.content:
            raise ValueError("No content blocks in response")

        tool_results = []

        for block in response.content:
            tool_use_block = self._get_tool_use_block(provider_type, block)
            if not tool_use_block:
                continue

            if provider_type == AgentProviderType.BEDROCK.value:
                tool_name = tool_use_block.get("name")
                tool_id = tool_use_block.get("toolUseId")
                input_data = tool_use_block.get("input", {})
            else:
                # Anthropic object-style block
                tool_name = tool_use_block.name
                tool_id = tool_use_block.id
                input_data = tool_use_block.input

            await self.callbacks.on_tool_start(
                tool_name, input_data, metadata={"agent_info": agent_info}
            )

            result = await self._call_mcp_tool(tool_name, input_data)

            await self.callbacks.on_tool_end(
                tool_name, input_data, result, metadata={"agent_info": agent_info}
            )

            # Only the text reaches the model; the structured data + widget ride on the ToolResult
            # for a UI-aware consumer (e.g. GroundedAgent), captured via on_tool_end above.
            model_text = result.content or json.dumps(result.structured_content, default=str)

            if provider_type == AgentProviderType.BEDROCK.value:
                tool_results.append(
                    {
                        "toolResult": {
                            "toolUseId": tool_id,
                            "content": [{"text": model_text}],
                        }
                    }
                )
            else:
                tool_results.append(
                    {
                        "type": "tool_result",
                        "tool_use_id": tool_id,
                        "content": model_text,
                    }
                )

        if provider_type == AgentProviderType.BEDROCK.value:
            return ConversationMessage(
                role=ParticipantRole.USER.value, content=tool_results
            )
        return {"role": ParticipantRole.USER.value, "content": tool_results}

    async def _call_mcp_tool(self, tool_name: str, input_data: dict) -> ToolResult:
        """Call a tool on the appropriate MCP server.

        Returns a :class:`~agent_squad.utils.tool.ToolResult` carrying the text (added to the model's
        context), the render-only ``structured_content``, and a ``UIPayload`` widget when the tool
        advertised one via its ``_meta.ui`` (fetched from the server as a resource)."""
        entry = self._tool_map.get(tool_name)
        if entry is None:
            return ToolResult(content=f"Tool '{tool_name}' not found in any connected MCP server")

        try:
            call_result = await entry.session.call_tool(tool_name, input_data)
        except Exception as exc:  # noqa: BLE001
            return ToolResult(content=f"Error calling tool '{tool_name}': {exc}")

        parts = [
            getattr(item, "text", str(item))
            for item in (getattr(call_result, "content", None) or [])
        ]
        text = "\n".join(parts)

        if getattr(call_result, "isError", False):
            # Surface the error text back to the model so it can react.
            return ToolResult(content=f"Tool error: {text}" if text else "Tool returned an error")

        structured = getattr(call_result, "structuredContent", None) or {}

        ui: Optional[UIPayload] = None
        if entry.ui:
            template = await self._template_for(entry.session, entry.ui)
            if template is not None:
                mime_type, body = template
                ui = UIPayload(
                    resource_uri=entry.ui,
                    mime_type=mime_type,
                    template=body,
                    structured_content=structured,
                    meta=getattr(call_result, "meta", None),
                )

        return ToolResult(content=text, structured_content=structured, ui=ui)

    async def _template_for(self, session: Any, resource_uri: str) -> Optional[tuple[str, str]]:
        """Fetch (and cache) a UI template resource by URI. Returns ``(mime_type, body)`` or ``None``.

        A resource arrives as text or as a base64 ``blob``; the blob is decoded as UTF-8 markup."""
        cache_key = (id(session), resource_uri)
        if cache_key in self._template_cache:
            return self._template_cache[cache_key]
        try:
            read_result = await session.read_resource(AnyUrl(resource_uri))
        except Exception:  # noqa: BLE001
            return None
        contents = getattr(read_result, "contents", None) or []
        if not contents:
            return None
        first = contents[0]
        mime_type = getattr(first, "mimeType", None) or "text/html;profile=mcp-app"
        body = getattr(first, "text", None)
        if body is None:
            blob = getattr(first, "blob", None)
            if isinstance(blob, str):
                try:
                    body = base64.b64decode(blob).decode("utf-8")
                except Exception:  # noqa: BLE001
                    body = None
        if body is None:
            return None
        self._template_cache[cache_key] = (mime_type, body)
        return (mime_type, body)

    # ------------------------------------------------------------------
    # Format conversion helpers
    # ------------------------------------------------------------------

    def to_bedrock_format(self) -> list[dict[str, Any]]:
        """Return MCP tools in the Bedrock ``toolSpec`` format.

        .. note::
            This is a *synchronous* method.  If the provider has not yet been
            connected you must call ``await provider._ensure_connected()``
            before using this method, or use the agent's async flow which will
            connect lazily via :meth:`tool_handler`.
        """
        result = []
        for tool_name, entry in self._tool_map.items():
            if not entry.model_visible:
                continue  # app-only tool: callable by the UI, never advertised to the model
            mcp_tool = entry.tool
            input_schema = (
                mcp_tool.inputSchema
                if isinstance(mcp_tool.inputSchema, dict)
                else {"type": "object", "properties": {}}
            )
            result.append(
                {
                    "toolSpec": {
                        "name": tool_name,
                        "description": mcp_tool.description or "",
                        "inputSchema": {"json": input_schema},
                    }
                }
            )
        return result

    def to_claude_format(self) -> list[dict[str, Any]]:
        """Return MCP tools in the Anthropic / Claude ``input_schema`` format."""
        result = []
        for tool_name, entry in self._tool_map.items():
            if not entry.model_visible:
                continue  # app-only tool: callable by the UI, never advertised to the model
            mcp_tool = entry.tool
            input_schema = (
                mcp_tool.inputSchema
                if isinstance(mcp_tool.inputSchema, dict)
                else {"type": "object", "properties": {}}
            )
            result.append(
                {
                    "name": tool_name,
                    "description": mcp_tool.description or "",
                    "input_schema": input_schema,
                }
            )
        return result

    def to_anthropic_format(self) -> list[dict[str, Any]]:
        """Alias for :meth:`to_claude_format` (same wire format)."""
        return self.to_claude_format()

    def to_openai_format(self) -> list[dict[str, Any]]:
        """Return MCP tools in the OpenAI function-calling format."""
        result = []
        for tool_name, entry in self._tool_map.items():
            if not entry.model_visible:
                continue  # app-only tool: callable by the UI, never advertised to the model
            mcp_tool = entry.tool
            input_schema = (
                mcp_tool.inputSchema
                if isinstance(mcp_tool.inputSchema, dict)
                else {"type": "object", "properties": {}}
            )
            # Ensure required field is present for strict mode compatibility
            parameters = {**input_schema}
            if "additionalProperties" not in parameters:
                parameters["additionalProperties"] = False
            result.append(
                {
                    "type": "function",
                    "function": {
                        "name": tool_name,
                        "description": mcp_tool.description or "",
                        "parameters": parameters,
                    },
                }
            )
        return result
